### ISSUE TITLE
retrieving db keys causes some keys to become invisible to subsequent lookups

### ISSUE THREAD

**Issue by lebedov** [coleifer/unqlite-python#28]
Using unqlite-python 0.5.3 with Python 2.7.11 on MacOS 10.11.6, the `print` statement after the second loop prints 488, not 500. This doesn't occur if I comment out `keys = [k for k in db]`.

```python
import unqlite
file_name = 'test.db'

db = unqlite.UnQLite(file_name)
for i in range(500):
    db['key_%s' % i] = 100*'x'
db.close()

db = unqlite.UnQLite(file_name)
keys = [k for k in db]
print len(keys)

count = 0
for k in db.keys():
    db[k]
    count += 1
print count
db.close()
```

**Comment by coleifer** [comment]
Have you tried writing a minimal C version to verify the problem is the Python bindings?

**Comment by lebedov** [comment]
Yes - seems like it might be a problem with the C library itself. [Posted it](https://github.com/symisc/unqlite/issues/41) on the unqlite issue tracker.

**Comment by coleifer** [comment]
Interesting... I tried the same experiment. I used 1000 records and after a close/reopen iterate (1000), iterate again ... **976**?? Wtf! Interestingly I'm double the number of records short that you are (12 -> 24).

Closing and re-opening I get my 1000 back, but I'm curious to see which keys are missing....so here are the ones that were missing:

{'key_270',
 'key_271',
 'key_371',
 'key_390',
 'key_416',
 'key_417',
 'key_490',
 'key_516',
 'key_517',
 'key_531',
 'key_533',
 'key_534',
 'key_535',
 'key_536',
 'key_921',
 'key_922',
 'key_923',
 'key_933',
 'key_940',
 'key_941',
 'key_950',
 'key_951',
 'key_960',
 'key_970'}

I tried this twice times and got the same keys missing each time.

My conclusion at the moment? Run the fuck away from UnQLite.

**Comment by coleifer** [comment]
I've added a failing test-case despite the project author's suggestion that I upgrade to the latest version it is present in the 1.1.6 source amalgamation in their GH repo.

I'll keep an eye on https://github.com/symisc/unqlite/issues/41 and update the bundled sources as soon as it's fixed upstream but from what I currently understand this is a library issue rather than a bindings issue.

**Comment by coleifer** [comment]
This is fixed in version 0.6.0. I dropped the amalgamation and am using the individual sources instead which contained the fix for the above issue.

**Comment by lebedov** [comment]
Thanks!
